File: sharetravelBackend/src/main/java/com/salesianostriana/dam/sharetravelBackend/trip/model/Trip.java
package com.salesianostriana.dam.sharetravelBackend.trip.model;

import com.fasterxml.jackson.annotation.JsonFormat;
import com.salesianostriana.dam.sharetravelBackend.reserve.model.Reserve;
import com.salesianostriana.dam.sharetravelBackend.user.model.Driver;
import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.UuidGenerator;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

import java.time.LocalDateTime;
import java.util.List;
import java.util.UUID;

@Entity
@EntityListeners(AuditingEntityListener.class)
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Trip {
    @Id
    @GeneratedValue(generator = "UUID")
    @UuidGenerator
    @Column(columnDefinition = "uuid")
    private UUID id;

    private String departurePlace;

    private String arrivalPlace;

    //@JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "dd/MM/yyyy HH:mm:ss")
    private LocalDateTime departureTime;

    private int estimatedDuration;//EN MINUTOS

    //@JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "dd/MM/yyyy HH:mm:ss")
    private LocalDateTime arrivalTime;

    private double price;

    private String tripDescription;

    @ManyToOne(optional = false, fetch = FetchType.LAZY)
    @JoinColumn(name = "driver_id", nullable = false)
    private Driver driver;

    @ToString.Exclude
    @OneToMany(cascade = CascadeType.ALL, orphanRemoval = true)
    @JoinColumn(name = "trip_id")
    private List<Reserve> reserves;

    public void calculateArrivalTime() {
        if (departureTime != null && estimatedDuration > 0) {
            this.arrivalTime = departureTime.plusMinutes(estimatedDuration);
        }
    }

}

File: sharetravel_frontend/lib/model/response/passenger_reserves_response/trip.dart
import 'dart:convert';

import 'driver.dart';

class Trip {
  String? id;
  String? departurePlace;
  String? arrivalPlace;
  String? departureTime;
  int? estimatedDuration;
  String? arrivalTime;
  int? price;
  String? tripDescription;
  Driver? driver;

  Trip({
    this.id,
    this.departurePlace,
    this.arrivalPlace,
    this.departureTime,
    this.estimatedDuration,
    this.arrivalTime,
    this.price,
    this.tripDescription,
    this.driver,
  });

  factory Trip.fromMap(Map<String, dynamic> data) => Trip(
        id: data['id'] as String?,
        departurePlace: data['departurePlace'] as String?,
        arrivalPlace: data['arrivalPlace'] as String?,
        departureTime: data['departureTime'] as String?,
        estimatedDuration: data['estimatedDuration'] as int?,
        arrivalTime: data['arrivalTime'] as String?,
        price: data['price'] as int?,
        tripDescription: data['tripDescription'] as String?,
        driver: data['driver'] == null
            ? null
            : Driver.fromMap(data['driver'] as Map<String, dynamic>),
      );

  Map<String, dynamic> toMap() => {
        'id': id,
        'departurePlace': departurePlace,
        'arrivalPlace': arrivalPlace,
        'departureTime': departureTime,
        'estimatedDuration': estimatedDuration,
        'arrivalTime': arrivalTime,
        'price': price,
        'tripDescription': tripDescription,
        'driver': driver?.toMap(),
      };

  /// `dart:convert`
  ///
  /// Parses the string and returns the resulting Json object as [Trip].
  factory Trip.fromJson(String data) {
    return Trip.fromMap(json.decode(data) as Map<String, dynamic>);
  }

  /// `dart:convert`
  ///
  /// Converts [Trip] to a JSON string.
  String toJson() => json.encode(toMap());
}
